L80-csharp-load-balancing.md

```diff
@@ -0,0 +1,296 @@
+Title
+----
+* Author(s): James Newton-King
+* Approver: jtattermusch and markdroth
+* Status: Draft
+* Implemented in: csharp
+* Last updated: 2021-05-19
+* Discussion at: <google group thread> (filled after thread exists)
+
+## Abstract
+
+grpc-dotnet currently has no load balancing support. A channel is always created for a single address.
+
+This documentation outlines an initial load balancing implementation for the grpc-dotnet client.
+
+## Background
+
+grpc-dotnet currently doesn't support any form of load balancing in the client. The changes outlined in this proposal will enable basic gRPC load balancing and some of its related features. These include:
+
+* Name resolving. There is built-in support for resolving from DNS, and a static resolver. Can create custom resolvers.
+* Load balancing strategies. Includes built-in support for pick first and round-robin load balancing. Can create custom load balancers.
+* Connectivity status at sub-channel and channel level
+* Wait for ready
+
+## Proposal
+
+![Diagram](L80_graphics/diagram.png)
+
+### Initialization
+
+Channels are created with:
+
+* An address
+* An optional service provider
+* An optional service config
+
+```csharp
+var channel = GrpcChannel.ForAddress(
+    "dns:///backend.default.svc.cluster.local",
+    new GrpcChannelOptions { ChannelCredentials = ChannelCredentials.Insecure });
+var client = new Greet.GreeterClient(channel);
+
+var response = await client.SayHelloAsync(new HelloRequest { Name = "world" });
+Console.WriteLine(response.Message);
+```
+
+* Resolver is created from address scheme.
+  * Existing logic allows for `http` and `https`. These schemes are special-cased to never use a resolver.
+  * A collection of resolver factories are fetched from the service provider. One is selected using the address scheme, and it will create a resolver. For example, `dns` will match to `DnsAddressResolverFactory`, creating a `DnsAddressResolver`.
+* Load balancer is created from the load balancer name in service config.
+  * A collection of load balancer factories are fetched from the service provider. One is selected using the configured name (if none specified, then `pick_first` is used), creating a load balancer.
+  * Load balancer is initialized with an `IChannelController` instance by the channel. The controller calls back to the channel to create sub-channels and update the picker.
+
+### Connecting
+
+* No network activity until `GrpcChannel.ConnectAsync` is called, or the first RPC is started.
+* On connect, the `ConnectionManager` starts the resolver and prompts it to run immediately.
+* Resolver pushes addresses and optional service config to `ConnectionManager`.
+* Load balancer is created.
+  * Resolver can return a service config. If `DisableServiceConfig` is not `true` then the resolver's service config is used to get load balancer policy.
+  * Channel service config option is used as a fallback.
```

Comment:
How is this passed in by the application?

Reply:
Sample added.